### PR TITLE
codec: avoid overflowing error

### PR DESCRIPTION
**Pull request by BusyJay**
grpc c core can only handle message that is not larger than 4GiB.

### CODE REVIEW

File: src/codec.rs [hunk 1]
```diff
@@ -31,16 +31,23 @@ pub mod pb_codec {
...
+        if cap <= u32::MAX as usize {
+            unsafe {
+                let bytes = buf.realloc(cap);
+                let raw_bytes = &mut *(bytes as *mut [std::mem::MaybeUninit<u8>] as *mut [u8]);
+                let mut s = CodedOutputStream::bytes(raw_bytes);
+                t.write_to_with_cached_sizes(&mut s).map_err(Into::into)
+            }
+        } else {
+            Err(Error::Codec(
+                format!("message is too large: {} > u32::MAX", cap).into(),
```
Comment (hunterlxt) [inline src/codec.rs:49]: Add comment to explain the limit is from somewhere in grpc core

File: src/codec.rs [hunk 2]
```diff
@@ -56,20 +63,27 @@ pub mod pb_codec {
...
+            unsafe {
+                let bytes = buf.realloc(size);
+                let mut b = &mut *(bytes as *mut [std::mem::MaybeUninit<u8>] as *mut [u8]);
+                msg.encode(&mut b)?;
+                debug_assert!(b.is_empty());
+            }
+            Ok(())
+        } else {
+            Err(Error::Codec(
+                format!("message is too large: {} > u32::MAX", size).into(),
```
Comment (hunterlxt) [inline src/codec.rs:85]: same as above

**Review by hunterlxt — COMMENTED** [review]
Most LGTM

File: src/codec.rs (hunk 1 above)
Comment (hunterlxt) [reply src/codec.rs:49]: refer to https://github.com/tikv/grpc-rs/pull/496/files#diff-ca14eabd5ab83f72548cd7e44eaea3f56a5d2808e3275c6b2a97622ebe5d1dd0R9-R11 ?

**Comment by JmPotato** [comment]
Any updates?

**Comment by ekexium** [comment]
Any updates?

**Comment by hunterlxt** [comment]
seem that CI should be fixed first

**Comment by BusyJay** [comment]
The problem is it's not enough to only include this check as rust-protobuf doesn't return correct length.

**Comment by BusyJay** [comment]
I have sent https://github.com/stepancheg/rust-protobuf/pull/571 to fix the panic. This PR can be merged first.

**Comment by BusyJay** [comment]
Nightly failure is a known issue: https://github.com/tokio-rs/prost/issues/526.

**Review by hicqu — APPROVED** [review]

**Comment by BusyJay** [comment]
The changed public APIs are all from codec. Those APIs are used by generated code, and those code can work with the new API without any changes. So I plan to land it without bumping minor versions to make it possible to fix overflowing error for previous versions.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**